### PR TITLE
Clipboard API for Plugins

### LINKED ISSUE #5519: [vscode] complete clipboard api
See https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/vscode.d.ts#L5977-L5993

### PR DESCRIPTION
**Pull request by azatsarynnyy**
#### What it does
Adds Clipboard API for Plugins.
Closes #5519 
This PR is based on the code extracted from #5527 

#### Some notes about the implementation
Different browsers and it's native extensions work a bit differently with the APIs around the clipboard: Async Clipboard API, Permissions API, document.execCommand API.
So the PR adds a minimal level of support for Chrome and FireFox browsers. Support for other cases may be added later as needed.

#### How to test
Run Theia with the test VS Code extension https://github.com/azatsarynnyy/test-clipboard-api/blob/master/clipboard-api-test-0.0.1.vsix

To test [`vscode.env.clipboard.readText` API](https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/vscode.d.ts#L5982-L5986):
- copy an arbitrary text to the clipboard
- call `Clipboard: read from clipboard` command
- the text from the clipboard should be shown with a notification

To test [`vscode.env.clipboard.writeText` API](https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/vscode.d.ts#L5988-L5992):
- select an arbitrary text in Theia editor
- call `Clipboard: write selection to clipboard` command
- paste the clipboard content to any place to check that it was really copied

#### Browser-specific behavior:
- Chrome
The user is able to block the access to the clipboard by setting a preference
perm

In that case, Theia shows the message
image

- FireFox
By default, Async Clipboard API is disabled but it can be enabled by setting `dom.events.testing.asyncClipboard` preference. User will be informed about that with a notification
image

Besides that, Async Clipboard API isn't gated with the permissions in FireFox, so it just bypassed.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
@azatsarynnyy is it ready for a review?

File: packages/core/src/browser/browser-clipboard-service.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2019 TypeFox and others.
+ * Copyright (C) 2019 RedHat and others.
```
Comment (akosyakov) [inline packages/core/src/browser/browser-clipboard-service.ts]: @theia-ide/ip-help I think it is fine, right? I've never merged these files or we should always preserver copyright even while copying code from each other PRs?

**Comment by azatsarynnyy** [comment]
@akosyakov it's not ready yet. I want to create another test plugin which will test Clipboard API only. I'll ping you once it's ready.

**Comment by azatsarynnyy** [comment]
It's ready for review and test

**Comment by akosyakov** [comment]
I'm offline tomorrow. If someone gives a try proposed test cases in the description on different browsers it would be helpful. Also if something does not work think about suggestion to fix them, i.e. use DOM api as we do in some places in the code for `paste`, for `copy` don't have an idea.

File: packages/core/src/browser/browser-clipboard-service.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+export class BrowserClipboardService implements ClipboardService {
+
+    async readText(): Promise<string> {
+        if ('permissions' in navigator && 'clipboard' in navigator) {
+            const result = await (<NavigatorPermissions>navigator['permissions']).query({ name: 'clipboard-read' });
+            if (result.state === 'granted' || result.state === 'prompt') {
+                return (<NavigatorClipboard>navigator['clipboard']).readText();
+            }
+        }
+        return '';
```
Comment (akosyakov) [inline packages/core/src/browser/browser-clipboard-service.ts]: We should fall back to the old strategy if a browser does not support `permissions` or `clipboard`, something like here: https://github.com/sindresorhus/copy-text-to-clipboard/blob/master/index.js

We can also just use `copy-text-to-clipboard` for this case. For `paste` we can try to install a temporary `paste` listener on the document, and trigged  `document.execCommand('paste')`.

**Comment by akosyakov** [comment]
> it did not seem to work on Firefox

It is strange. Latest Firefox supports should support these APIs. @fangnx which version of Firefox do you use? 

@azatsarynnyy could you debug please what is going wrong on FireFox. Theia claims to support latest version of Firefox and Chrome.

**Comment by akosyakov** [comment]
We need to verify as well that @benoitf use case is working: https://github.com/theia-ide/theia/pull/5527#pullrequestreview-253983347

**Comment by fangnx** [comment]
> > it did not seem to work on Firefox
> 
> It is strange. Latest Firefox supports should support these APIs. @fangnx which version of Firefox do you use?
> 
> @azatsarynnyy could you debug please what is going wrong on FireFox. Theia claims to support latest version of Firefox and Chrome.

Firefox Quantum 68.0.1 (64-bit), Ubuntu 18.04.1

**Comment by akosyakov** [comment]
From https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#Browser_compatibility:
> Firefox only supports reading the clipboard in browser extensions, using the "clipboardRead" extension permission.

> Writing to the clipboard is available without permission in secure contexts and browser extensions, but only from user-initiated event callbacks. Browser extensions with the "clipboardWrite" permission can write to the clipboard at any time.

So basically FireFox does not support these APIs yet :( We need a fallback strategy.

**Comment by akosyakov** [comment]
Also found in https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read:

> At this time, while Firefox does implement read(), it does not recognize the "clipboard-read" permission, so attempting to use the Permissions API to manage access to the API will not work.

So maybe we should no `query` permissions, but just always call `writeText` or `readText`. As far as I understood they will request permissions automatically and if not possible throw. If they throw then we should fallback to document based strategy.

**Comment by azatsarynnyy** [comment]
Thank you for testing it, guys! :bowing_man: 
I'll be able to continue to work on your feedback on Tuesday.

**Comment by azatsarynnyy** [comment]
It doesn't work in FF for me also. I'm going to investigate the problem.

**Comment by azatsarynnyy** [comment]
> Also found in https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read:
> 
> > At this time, while Firefox does implement read(), it does not recognize the "clipboard-read" permission, so attempting to use the Permissions API to manage access to the API will not work.
> 
> So maybe we should no `query` permissions, but just always call `writeText` or `readText`. As far as I understood they will request permissions automatically and if not possible throw. If they throw then we should fallback to document based strategy.

Async Clipboard API works well in Chrome without asking the permissions.
FF still requires enabling the `dom.events.testing.asyncClipboard` preference in `about:config` to turn on that API.

**Comment by akosyakov** [comment]
> Async Clipboard API works well in Chrome without asking the permissions.
FF still requires enabling the dom.events.testing.asyncClipboard preference in about:config to turn on that API.

@azatsarynnyy Can we somehow detect it and notify a user to enable it?

**Comment by azatsarynnyy** [comment]
I think relying on `dom.events.testing.asyncClipboard` preference would be a hack since it's actually intended to be used by FF automated tests only. It allows bypassing the security checks when interacting with the clipboard. See https://github.com/mozilla/gecko-dev/blob/e04021f29e6d8a37753ba2b510432315ce05a8d7/dom/events/Clipboard.h#L61

Although FF declares support of Async Clipboard API, it's exposed to secure contexts and browser extension only. So, looks like we cannot use it for this task.

`execCommand` API can be used only in a short running user-generated event handlers, thus it also doesn't suit us.

As I understand, the only way we can follow is by relying on some 3rd party lib.
WDYT?

**Comment by akosyakov** [comment]
> As I understand, the only way we can follow is by relying on some 3rd party lib.

I don't think that 3rd party lib will do anything special, just calling `execCommand`.

> execCommand API can be used only in a short running user-generated event handlers, thus it also doesn't suit us.

Could we try it anyway, like suggested in https://github.com/theia-ide/theia/pull/5994#discussion_r317029396? We are already using such approach in some places it was good so far. It should be only a fallback.

**Comment by azatsarynnyy** [comment]
@akosyakov the reason why we cannot use `document.execCommand` API as a fallback is it works without any restrictions in Chrome but not in FireFox. And `copy-text-to-clipboard` doesn't work in FF for the same reason.

The main issue here is that FF requires calling `document.execCommand('cut'/'copy')` APIs inside the short-living event handlers of user actions like mouse click, keypress etc.

I examined all the places in Theia where `document.execCommand` API is used. All of them are actually called as Theia commands handlers, which means they're called inside the short-living event handlers of user actions, e.g. click, keypress.

I tried to call `ClipboardService.writeText` as Theia command handler, with the following code:
```javascript
       const onCopy = (e: ClipboardEvent) => {
            document.removeEventListener('copy', onCopy);
            if (e.clipboardData) {
                e.clipboardData.setData('text/plain', value);
                e.preventDefault();
            }
        };
        document.addEventListener('copy', onCopy);
        document.execCommand('copy');
```
It worked well in both, Chrome and FF.

But when I call it through the Plugin API, it works in Chrome only. And FireFox hints in the console with the following message:
`document.execCommand(‘cut’/‘copy’) was denied because it was not called from inside a short running user-generated event handler.`

I believe it's because of `document.execCommand` API is called by plugin via JSON-RPC but not inside a short-living event handler of user action.

**Comment by azatsarynnyy** [comment]
To allow the user using Clipboard Plugin API in FireFox, I don't see any other options except recommending to enable `dom.events.testing.asyncClipboard` preference.

**Comment by akosyakov** [comment]
> To allow the user using Clipboard Plugin API in FireFox, I don't see any other options except recommending to enable dom.events.testing.asyncClipboard preference.

@azatsarynnyy Let's notify a user with an error that he head to enable it and how in FF. Do I understand correctly that then `readTest` and `writeText` should work as well in Firefox and we don't need to work around them?

**Comment by akosyakov** [comment]
Thought a bit about it again, we still want to have a fallback for native extensions, since they can call `ClipboardService` from the event handlers and then will work better. But it can be done by a separate PR, for this one notifying a user that action failed and explain how to fix it would be enough.

**Comment by azatsarynnyy** [comment]
> Do I understand correctly that then readTest and writeText should work as well in Firefox and we don't need to work around them?

correct

**Comment by akosyakov** [comment]
@azatsarynnyy please merge it

**Comment by azatsarynnyy** [comment]
The follow-up issue https://github.com/eclipse-theia/theia/issues/6209

**Pull request merged**